pkg/oci/build_metadata.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/cilium/ebpf"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/run/types"
+)
+
+func loadSpec(progContent []byte) (*ebpf.CollectionSpec, error) {
+	progReader := bytes.NewReader(progContent)
+	spec, err := ebpf.LoadCollectionSpecFromReader(progReader)
+	if err != nil {
+		return nil, fmt.Errorf("loading spec: %w", err)
+	}
+	return spec, err
+}
+
+// getAnySpec returns the spec for any architecture found. It's used to generate the metadata so we
+// don't care about the architecture this file was generated for.
+func getAnySpec(opts *BuildGadgetImageOpts) (*ebpf.CollectionSpec, error) {
+	var progPath string
+
+	// TODO: we could perform a sanity check to be sure different architectures generate the
+	// same metadata, but that's too much for now.
+	for _, path := range opts.EBPFObjectPaths {
+		progPath = path
+		break
+	}
+
+	if progPath == "" {
+		return nil, fmt.Errorf("no eBPF object file found")
+	}
```

Comment:
The suggestion doesn't work as `opts.EBPFObjectPaths` is a `map[string]string`. 

> Isn't there a risk of getting any spec? Let's imagine a case where the user creates a gadget for both architecture but which gets a specific field for a given architecture. Is this case realistic? If so, should we get the spec corresponding to the host architecture and defaulting to it?

All of this logic is done at build time, it's for generating a metadata skeleton from the built objects. I think event structures, maps definition, etc. should be the same for all architectures. If there is a field only available for a given architecture, it should be empty for the architecture that doesn't support it, otherwise it'll big a mess to support that from our side.

In the extreme case those things are not the same for different architectures, the validation at runtime will fail, so I think we're safe there. Added a comment to highlight this on the code.

Reply:
> Anyway to get an iterator on the map and just dereference it?

I'm not sure, the way I used seems to be the simplest to get the single element from the map.